src/main/java/com/google/devtools/build/lib/rules/cpp/CcLibraryHelper.java

```diff
@@ -1367,6 +1368,72 @@ private CppCompilationContext initializeCppCompilationContext(CppModel model) {
       }
     }
 
+    if (ruleContext.getFragment(CppConfiguration.class).experimentalEnableImplicitHeaderMaps()) {
+      ImmutableList.Builder<Artifact> headerMapsBuilder = ImmutableList.builder();
+      String targetName = ruleContext.getTarget().getName();
+
+      HeaderMapInfo.Builder internalHeaderMapInfo = new HeaderMapInfo.Builder();
+      internalHeaderMapInfo.addHeaders(publicHeaders.getHeaders());
+      internalHeaderMapInfo.addHeaders(privateHeaders);
+      internalHeaderMapInfo.addHeaders(publicTextualHeaders);
+      Artifact internalHeaderMap = ruleContext.getPackageRelativeArtifact(PathFragment.create(targetName + "_internal.hmap"),
+                ruleContext
+                .getConfiguration()
+                .getGenfilesDirectory(ruleContext.getRule().getRepository()));
+      ruleContext.registerAction(
+        new HeaderMapAction(ruleContext.getActionOwner(),
+            internalHeaderMapInfo.build().getSources(),
+            internalHeaderMap));
+      contextBuilder.addQuoteIncludeDir(internalHeaderMap.getExecPath());
+      headerMapsBuilder.add(internalHeaderMap);
+
+      String namespace;
+      if (ruleContext.attributes().has("header_namespace")) {
+         namespace = ruleContext.attributes().get("header_namespace", Type.STRING);
+      } else {
+         namespace = ruleContext.getRule().getName();
+      }
+
+      // Construct the dep headermap.
+      // This header map additionally contains namespaced headers so that a user
+      // can import headers of the form Namespace/Header.h from headers within
+      // the current target.
+      HeaderMapInfo.Builder depHeaderMapInfo = new HeaderMapInfo.Builder();
+      depHeaderMapInfo.setNamespace(namespace);
+      depHeaderMapInfo.addNamespacedHeaders(publicHeaders.getHeaders());
+      depHeaderMapInfo.addHeaders(publicHeaders.getHeaders());
+      depHeaderMapInfo.addNamespacedHeaders(privateHeaders);
+      depHeaderMapInfo.addHeaders(privateHeaders);
+      depHeaderMapInfo.addNamespacedHeaders(publicTextualHeaders);
+      depHeaderMapInfo.addHeaders(publicTextualHeaders);
+
+      // Merge all of the header map info from deps. The headers within a given
+      // target have precedence over over dep headers ( See
```

Comment:
SuperNit: extra space between ( and See.

Reply:
Will Fix